File: typescript/Behavioral/18-Observer/Subject/JobCard.ts
import type { IJobCardObserver } from "../Observer/IJobCardObserver";

export class JobCard {
  readonly jobId: string = Math.random().toString(36).slice(2, 10);

  private _status: string = "Draft";
  private _assignee: string | undefined;
  private _notes: string | undefined;
  private _cost: number;
  private _observers: IJobCardObserver[] = [];

  constructor(
    readonly title: string,
    readonly category: string,
    cost: number,
  ) {
    this._cost = cost;
  }

  get status(): string {
    return this._status;
  }
  get assignee(): string | undefined {
    return this._assignee;
  }
  get notes(): string | undefined {
    return this._notes;
  }
  get cost(): number {
    return this._cost;
  }

  subscribe(observer: IJobCardObserver): void {
    this._observers.push(observer);
  }

  unsubscribe(observer: IJobCardObserver): void {
    this._observers = this._observers.filter((o) => o !== observer);
  }

  notifyObservers(eventName: string): void {
    for (const observer of this._observers) {
      observer.update(eventName, this);
    }
  }

  assign(handyman: string): void {
    this._assignee = handyman;
    this._status = "Assigned";
    this.notifyObservers("JobAssigned");
  }

  startWork(): void {
    this._status = "InProgress";
    this.notifyObservers("JobStarted");
  }

  pauseWork(reason: string): void {
    this._status = "Paused";
    this._notes = reason;
    this.notifyObservers("JobPaused");
  }

  complete(): void {
    this._status = "Completed";
    this._notes = undefined;
    this.notifyObservers("JobCompleted");
  }

  toString(): string {
    return [
      `Job #${this.jobId} '${this.title}' [${this.category}]`,
      `  Status  : ${this._status}`,
      `  Assignee: ${this._assignee ?? "none"}`,
      `  Notes   : ${this._notes ?? "none"}`,
      `  Cost    : N$${this._cost.toFixed(2)}`,
    ].join("\n");
  }
}

File: csharp/Behavioral/18-Observer/Subject/JobCard.cs
namespace Observer;

public class JobCard : IJobCardSubject
{
    public string JobId { get; } = Guid.NewGuid().ToString("N")[..8];

    private string _status = "Draft";
    private string? _assignee;
    private string? _notes;
    private decimal _cost;

    public string Title { get; }
    public string Category { get; }
    public string Status => _status;
    public string? Assignee => _assignee;
    public string? Notes => _notes;
    public decimal Cost => _cost;

    private readonly List<IJobCardObserver> _observers = new();

    public JobCard(string title, string category, decimal cost)
    {
        Title = title;
        Category = category;
        _cost = cost;
    }

    public void Subscribe(IJobCardObserver observer) => _observers.Add(observer);
    public void Unsubscribe(IJobCardObserver observer) => _observers.Remove(observer);

    public void NotifyObservers(string eventName)
    {
        foreach (var observer in _observers)
            observer.Update(eventName, this);
    }

    public void Assign(string handyman)
    {
        _assignee = handyman;
        _status = "Assigned";
        NotifyObservers("JobAssigned");
    }

    public void StartWork()
    {
        _status = "InProgress";
        NotifyObservers("JobStarted");
    }

    public void PauseWork(string reason)
    {
        _status = "Paused";
        _notes = reason;
        NotifyObservers("JobPaused");
    }

    public void Complete()
    {
        _status = "Completed";
        _notes = null;
        NotifyObservers("JobCompleted");
    }

    public override string ToString() =>
        $"Job #{JobId} '{Title}' [{Category}]\n" +
        $"  Status  : {_status}\n" +
        $"  Assignee: {_assignee ?? "none"}\n" +
        $"  Notes   : {_notes ?? "none"}\n" +
        $"  Cost    : N${_cost:F2}";
}
